core/src/main/java/org/apache/accumulo/fate/zookeeper/ZooCache.java

```diff
@@ -155,8 +167,22 @@ public void process(WatchedEvent event) {
       }
 
       switch (event.getType()) {
-        case NodeDataChanged:
         case NodeChildrenChanged:
+          if (event.getPath().endsWith("conf")) {
+            Matcher confDirMatcher = TABLE_CONFIG_DIR_PATTERN.matcher(event.getPath());
+            if (confDirMatcher.matches()) {
+              try {
+                clear(event.getPath());
+                getZooKeeper().exists(event.getPath(), watcher);
```

Comment:
I would contend that it sets a data watch.  But the main idea here is that this event is triggered on a 'conf' directory in the first place and if  the watch on that znode is not reset it will not be triggered again which is why @ivakegg  and @keith-turner may have seen table config persistence anomalies over time.   Accumulo sets the ZCacheWatcher watcher member object in ZooCache to watch on the 'conf' directories outside of this case statement.  All I am doing here is actually taking a logical, reasonable and productive action when that occurs.  On master branch all that happens is that the value (null) associated with the 'conf' directory is deleted from ZooCache.  By doing that we are totally missing the opportunity to 'refresh' the configs for that table.  In the event that for some reason there is no watch yet on a table config then we will not get the event that it has been added or changed.

So to recap.  I am just taking a course of action here that I think is required and should be done and resetting the watch on that zoopath for the tables conf directory so the event will be retriggered if a child changes again.  Remember.  We are not setting the first watch here.  Its being done elsewhere for a good reason and we just drop the resulting event on the floor IMHO.

Reply:
According to the zookeeper documentation, it will set a data watch.  A quick look through the code it looks like the data is for the `...\conf\` node is initialized with `new byte[0]` and I don't see it set to anything else.  Is there somewhere that I'm missing that sets / uses the data directly on the `...\conf` node?  Or, is the `...\conf` node a parent node for the rest of the configuration?  If the `...\conf` node exists solely so that child nodes can be created / contained, what value is there in setting a data watch for a change?

Reply:
What I think @EdColeman is that the intention was to do what I am trying to do now.  We can't just do nothing when this event occurs.  It is a great opportunity to refresh the table configs for an individual table if the NodeChildrenChanged event for the table 'conf' directories can always reliably be captured.  Thanks for looking into the documentation.

Reply:
To (hopefully) clarify my basic question - where in the code does something cause a data watch event to be triggered on the `...\conf` node?  What data is stored on the node, what uses it, and where is it modified after the initial creation?

Reply:
When some function like TablePropUtil.setTableProerty or TablePropUtil.deleteTableProperty is called ( or ZooUtil.putData during cloning) on a table config zoopath.  ZooKeeper will send a watch event of type NodeChildrenChange for the 'conf' zoopath to watcher in ZooCache and it will be processed in ZCacheWatcher.process().   Watchers are one-time triggers and need to be reset.  That is what this PR is doing now.

Reply:
Right now, I'm not asking about child change events.  For the `exists(path, watch)` to fire a data change event, something needs to put data to that node - child change events are a different situation and are not going to be updated with the data watch placed in a zookeeper exists call.

Reply:
So what you are saying is that I need to call getZookeeper().getChildren(zpath, wathcher) to get the NodeChildrenChanged event when the children change.  I will give that a try @EdColeman .  Thanks for the information.

Reply:
@EdColeman .  What I am seeing during run time is that when a new table config parameter is set on a table (in the Accumulo shell) the NodeChildrenChanged watch event for the parameters parent 'conf' directory gets triggered for 'exist watchers'.  Setting a watch on the new table config parameter can be done in the case statement too along with resetting the watcher on the new child nodes as you mentioned above.   

This part of the PR really isn't the main part.  It really just looks like the safe and conservative thing to do. 

The main idea of the PR is to not create watches on znodes that don't exist yet and may never come into existence.  Only create watches on table configs actually in ZooKeeper.  By doing this the delete operation on the table actually deletes all the table configs zoonodes and the corresponding watches go away also.  This keeps ZooKeeper from expanding the watcher maps in the ZooKeeper class from growing to the point of making ZooKeeper run out of memory and crash.  

If this modification to the case statement is considered undesirable and without any merit then I can remove it I am pretty sure.  I just think it is a good way to ensure the cache is consistent with what is actually in ZooKeeper.

Reply:
At this point, I'm trying to make sure that my understanding of zookeeper and my reading of the zookeeper documentation is correct so that these changes can be evaluated for correctness  
The original question that was asked is "if the watch should be added with an exits() or a getChildren() call". You maintain that this is the safe and conservative thing to do.  With my understanding of zookeeper, the watcher set by exists in this case is unnecessary.

Specifically, looking at [zookeeper user's guide - semantics of watches](http://zookeeper.apache.org/doc/r3.4.14/zookeeperProgrammers.html#ch_zkWatches) 

> - Created event: Enabled with a call to exists.
> - Deleted event: Enabled with a call to exists, getData, and getChildren.
> - Changed event: Enabled with a call to exists and getData.
> - Child event: Enabled with a call to getChildren.

This leads me to believe that  child events (node created, deleted, modified) are not enabled / called with an exists() call.  Using zookeeper directly, I performed the following:
```
created node /a
call exists(/a, true)  - this sets the exists watcher.
created node /a/b    (the exists watcher was not triggered)
call getChildren(/a, true) - this sets a second watcher for child events.
created node /a/c  - the child event watcher is triggered.
created node /a/d - no watchers triggered (watcher was not reset after /a/c create event)
modified the data on /a - the data watcher set with exists is triggered - also showing that it was not cleared with the child modification
```
This shows me that the exists() watcher is not triggered / cleared on child events - it is triggered when the data of node /a is modified.

You state that you see the exits watcher being triggered.  To me that implies something changed the data on the node.  Where in the code (other than initialization) is something changing the data?  Where in the code does anything read the data from the config node? (that would imply that it is being set / used). All I see is it being set at initialization (and on clone) to new byte[0].  Or, can you demonstrate a zookeeper call sequence that replicates what you are seeing?

My suspicion is the NodeChildrenChanged is being set with a call to getChildren() somewhere and that is what is being triggered, not with an exists() call.

I'm working an IT example to see if that adds clarity to zookeeper usage in the system beyond a toy zookeeper example directly.

Reply:
I am doing my testing and evaluation inside of the Accumulo shell in an UNO instance.  I tail my Accumulo logs with this command looking for my INFO debug statement:
tail -f install/logs/accumulo/* | grep 'NodeChildrenChanged: resetting watcher for'

Adding or deleting a table config will trigger the NodeChildrenChanged event to the 'conf' znode.  Do it in the Accumulo shell.  Try the steps below.

I repeatedly hit the NodeChildrenChanged case when I add a table config parameter to a table or delete a config parameter with 'config -t table1 -s table.bloom.enabled=true' or any other config that is not a default parameter.  Like I said,  I do this all in the Accumulo shell.  I use the zkCli.sh script to inspect the cversion and dataVersion fields with commands like 'get /accumulo/a328a8a6-78d6-4725-9bfc-80e5ec3a8639/tables/2/conf'. The cversion field changes the dataVersion doesn't.  I am reentering the NodeChildrenChanged case statement over and over because a reset the exists watcher on the node.  Watchers are one-time triggers and you have to reset them.  We were not doing that before.  Like I said before,  these lines of code are not central to this PR.  They just might help out though.  I ran the code with getZookeeper().get Children( zpath, watcher) to.  That did not harm anything.

The real evaluation of this PR is just verifying the removal of watches with the 'whcp' command I listed above - IMHO.  I can roll back to the old switch statement if these lines are causing confusion and they can be evaluated in the future.  I like them though.

We can do some pair programming at the Hack session next Tuesday and I can show you.

Reply:
The most important thing I think here is that when the watch is re-set here, it re-sets the same watches that were triggered. If `exists` does that, that's fine. But, since we'd be using the parent node to watch for changes in the children (instead of watching directly for non-existent child node changes), it might need to use `getChildren` to ensure the watch is set to look for child node changes.

Much of this complexity goes away entirely if we were to transition from storing the config as separate child nodes of `conf/` and serialize the configuration directly into the data of `conf/` as a JSON blob. I think that is probably a more worthy pursuit than trying to fix this in ZooCache. I'm going to look into doing that in a separate PR. I've looked at it in the past, and I think I have a good idea of what needs to be done. It should be easier now than when I looked at it in the past, since observers were removed and the upgrade code was improved.

Reply:
The watches on the children nodes of the 'conf' directories will be set as data watches the first time that ZooCache.get is called on the individual zoopaths of those configs.  All of the table configurations will get watched eventually like before if they really exist in zookeeper.  I agree that calling getChildren seems like a wise idea but all you are doing is putting a watch on a new child that is not in the ZooCache yet anyways and will only get into it when ZooCache.get is finally called on that (child) table configurations zoopath.   @keith-turner  made a great point to me that we should be limiting calls to zookeeper inside of the process method.